protos/test_result.proto

```diff
@@ -0,0 +1,50 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package wptdashboard;
+
+import "bq_table_name.proto";
+import "browser.proto";
+import "engine.proto";
+import "google/protobuf/timestamp.proto";
+import "operating_system.proto";
+import "sub_test_status.proto";
+import "test_status.proto";
+
+// An result for a single test or sub-test from a TestRun.
+message TestResult {
+  option (gen_bq_schema.table_name) = "test_results";
+
+  OperatingSystem os = 1;
+  string os_version_str = 2;
+
+  Engine engine = 3;
+  Browser browser = 4;
+  string browser_version_str = 5;
+
+  google.protobuf.Timestamp test_run_start_time = 6;
+  google.protobuf.Timestamp test_run_end_time = 7;
+  string long_wpt_hash = 8;
+  google.protobuf.Timestamp wpt_commit_time = 9;
+
+  string test_name = 10;
+  TestStatus status = 11;
+  string result_message = 12;
+
+  string test_sub_name = 13;
+  SubTestStatus sub_status = 14;
```

Comment:
Presently, the individual test result files that contain the "subtests" data produce JSON with the following shape: 

```json
{
  "test": "/shadow-dom/event-post-dispatch.html",
  "status": "ERROR",
  "message": "Object doesn't support property or method 'attachShadow'",
  "subtests": [
    {
      "status": "FAIL",
      "message": "assert_equals: expected 0 but got 3",
      "name": "Event properties post dispatch without ShadowRoots (composed: true)."
    }, {
      "status": "FAIL",
      "message": "assert_equals: expected 0 but got 3",
      "name": "Event properties post dispatch without ShadowRoots (composed: false)."
    }
  ]
}
```
(That's a real example https://storage.googleapis.com/wptd/5d55258739/edge-15-windows-10-sauce/shadow-dom/event-post-dispatch.html)

If I understand correctly, based on the design of `TestResult`, that data would look something like this: 


| os | os_version_str | engine | browser | browser_version_str | test_run_start_time | test_run_end_time | long_wpt_hash | wpt_commit_time | test_name | status | result_message | test_sub_name | sub_status | result_sub_message |
| -- | -------------- | ------ | ------- | ------------------- | ------------------- | ----------------- | ------------- | --------------- | --------- | ------ | -------------- | ------------- | ---------- | ------------------ |
| 3 | "10" | 2 | 2 | "15" |  test_run_start_time |  test_run_end_time | 5d552587396f6ee3148c4991a87130fbbdffd324 | wpt_commit_time | "/shadow-dom/event-post-dispatch.html" | 2 | "Object doesn't support property or method 'attachShadow'" | "Event properties post dispatch without ShadowRoots (composed: true)." | 2 | "assert_equals: expected 0 but got 3" |
| 3 | "10" | 2 | 2 | "15" |  test_run_start_time |  test_run_end_time | 5d552587396f6ee3148c4991a87130fbbdffd324 | wpt_commit_time | "/shadow-dom/event-post-dispatch.html" | 2 | "Object doesn't support property or method 'attachShadow'" | "Event properties post dispatch without ShadowRoots (composed: false)." | 2 | "assert_equals: expected 0 but got 3" |


Which now has duplicated the following information in both rows:

```js
  "test": "/shadow-dom/event-post-dispatch.html",
  "status": "ERROR",
  "message": "Object doesn't support property or method 'attachShadow'",
```

Reply:
Your assessment is exactly correct. The intent is that results can be filtered and/or grouped easily with a non-nested schema in BigQuery (and duplicate data is cheap).

Reply:
Put another way: We trade "wasted" storage for query simplicity (i.e., no need to join against a separate `sub_tests` table, or query over a `RECORD`/`STRUCT` field in BigQuery).